### PR TITLE
ARROW-11697: [Rust][DataFusion] Add provider for user defined function

### PR DESCRIPTION
**Pull request by wqc200**
select database();
we want to get database name in function database(), so add a provider for user defined function.

### CODE REVIEW

**Comment by alamb** [comment]
FYI the integration test failures are likely not due to the changes in this PR. The Dev/Lint failures probably need to be looked at. 

I will review this PR more carefully tomorrow

File: rust/arrow-flight/src/arrow.flight.protocol.rs [hunk 1]
```diff
@@ -499,8 +499,9 @@ pub mod flight_service_server {
     #[async_trait]
     pub trait FlightService: Send + Sync + 'static {
         #[doc = "Server streaming response type for the Handshake method."]
-        type HandshakeStream: Stream<Item = Result<super::HandshakeResponse, tonic::Status>>
```
Comment (alamb) [inline rust/arrow-flight/src/arrow.flight.protocol.rs:502]: I wonder if you meant to make this change in this PR?

This sometimes happens to me (I think it is something about using nightly vs stable `rustfmt`)

File: rust/datafusion/tests/func_provider.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+impl FuncProvider for DatabaseExec {
+    fn evaluate(&self, inputs: &[ColumnarValue]) -> Result<ColumnarValue> {
+        Ok(ColumnarValue::Scalar(ScalarValue::Utf8(Some(
+            self.db_name.to_string(),
+        ))))
+    }
+}
+
+async fn assert_provider_row_count(expected_count: String) -> Result<()> {
```
Comment (alamb) [inline rust/datafusion/tests/func_provider.rs:53]: the `expected_count` name here is confusing to me -- as it is a string. Perhaps `expected_name`?

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @wqc200  !

I will admit I don't fully understand the usecase for the user defined function provider. 

I think you could register a user defined function called `database()` that returned a specific value. For example, here is your test written using existing DataFusion functionality.

(as an aside,  this exerise took me longer to put together than I would have expected -- perhaps some better examples / helpers are in order):

```
async fn assert_provider_row_count(expected_name: String) -> Result<()> {
    let mut ctx = ExecutionContext::new();

    // implementation of `database()` function that returns expected_name
    let captured_name = expected_name.clone();
    let database_function = move |_args: &[ArrayRef]| {
        let res = StringArray::from(vec![Some(captured_name.as_str())]);
        Ok(Arc::new(res) as ArrayRef)
    };

    ctx.register_udf(create_udf(
        "database", // fuenction name
        vec![], // input argument types
        Arc::new(DataType::Utf8), // output type
        make_scalar_function(database_function)) // function implementation
    );

    let sql_results = ctx.sql(&format!("select database()"))?.collect().await?;

    let sql_result_col = array_value_to_string(sql_results[0].column(0), 0).unwrap();
    assert_eq!(sql_result_col, expected_name);

    Ok(())
}
```

If you want to avoid registering a new function for each possible database name, there is also the `scalar variable` functionality (e.g. you could do `select @@database@@ ...`. An example of such use is here:

https://github.com/apache/arrow/blob/bca7d2fe84ccd8fc1129cb4d85448eb0779c52c3/rust/datafusion/src/execution/context.rs#L716-L741

There may be some subtlety I am missing here that needs something like a `ActualFunction` but I don't understand it yet

**Comment by wqc200** [comment]
@alamb 

I'm using it in struct

```
use std::sync::Arc;

use arrow::array::ArrayRef;
use arrow::array::StringArray;
use arrow::datatypes::DataType;
use arrow::util::display::array_value_to_string;

use datafusion::error::Result;
use datafusion::execution::context::ExecutionContext;
use datafusion::logical_plan::create_udf;
use datafusion::physical_plan::functions::ActualFunction;
use datafusion::physical_plan::functions::make_scalar_function;

#[derive(Debug, Clone)]
pub struct TestProvider {
    db_name: String,
}

impl TestProvider {
    pub fn try_new(db_name: &str) -> Result<Self> {
        Ok(Self {
            db_name: db_name.to_string(),
        })
    }

    pub async fn test(&self, expected_name: String) -> Result<()> {
        let mut ctx = ExecutionContext::new();

        // implementation of `database()` function that returns expected_name
        let database_function = move |_args: &[ArrayRef]| {
            let res = StringArray::from(vec![Some(self.db_name.as_str())]);
            Ok(Arc::new(res) as ArrayRef)
        };

        ctx.register_udf(create_udf(
            "database", // fuenction name
            vec![], // input argument types
            Arc::new(DataType::Utf8), // output type
            ActualFunction::FuncImpl(make_scalar_function(database_function))), // function implementation
        );

        let sql_results = ctx.sql(&format!("select database()"))?.collect().await?;

        let sql_result_col = array_value_to_string(sql_results[0].column(0), 0).unwrap();
        assert_eq!(sql_result_col, expected_name);

        Ok(())
    }
}

#[tokio::test]
async fn test_func_provider_results() -> Result<()> {
    let p = TestProvider::try_new("test")?;
    p.test("test".to_string());
    Ok(())
}
```

but there's an error

> error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
>   --> datafusion/tests/func_provider.rs:43:23
>    |
> 43 |     pub async fn test(&self, expected_name: String) -> Result<()> {
>    |                       ^^^^^
>    |                       |
>    |                       this data with an anonymous lifetime `'_`...
>    |                       ...is captured here...
> ...
> 56 |             ActualFunction::FuncImpl(make_scalar_function(database_function))), // function implementation
>    |                                      -------------------- ...and is required to live as long as `'static` here

So i want to add a provider.

File: rust/arrow-flight/src/arrow.flight.protocol.rs (hunk 1 above)
Comment (wqc200) [reply rust/arrow-flight/src/arrow.flight.protocol.rs:502]: SELECT @@variable and SELECT DATABASE () are two different syntaxes.

File: rust/datafusion/tests/func_provider.rs (hunk 2 above)
Comment (wqc200) [reply rust/datafusion/tests/func_provider.rs:53]: yes, expected_name is right

File: rust/arrow-flight/src/arrow.flight.protocol.rs (hunk 1 above)
Comment (wqc200) [reply rust/arrow-flight/src/arrow.flight.protocol.rs:502]: SELECT @@variable and SELECT DATABASE () are two different syntaxes.

**Comment by alamb** [comment]
Hi @wqc200  -- I think the compiler issue in your example is due to the reference of `self` in the wrapped closure.

If you pass in a `clone` of the string, then the example compiles and passes:


```rust

use std::sync::Arc;

use arrow::array::ArrayRef;
use arrow::array::StringArray;
use arrow::datatypes::DataType;
use arrow::util::display::array_value_to_string;

use datafusion::error::Result;
use datafusion::execution::context::ExecutionContext;
use datafusion::logical_plan::create_udf;
use datafusion::physical_plan::functions::ActualFunction;
use datafusion::physical_plan::functions::make_scalar_function;

#[derive(Debug, Clone)]
pub struct TestProvider {
    db_name: String,
}

impl TestProvider {
    pub fn try_new(db_name: &str) -> Result<Self> {
        Ok(Self {
            db_name: db_name.to_string(),
        })
    }

    pub async fn test(&self, expected_name: String) -> Result<()> {
        let mut ctx = ExecutionContext::new();

        // implementation of `database()` function that returns expected_name
        let captured_name = self.db_name.clone();
        let database_function = move |_args: &[ArrayRef]| {
            let res = StringArray::from(vec![Some(captured_name.as_str())]);
            Ok(Arc::new(res) as ArrayRef)
        };

        ctx.register_udf(create_udf(
            "database", // fuenction name
            vec![], // input argument types
            Arc::new(DataType::Utf8), // output type
            ActualFunction::FuncImpl(make_scalar_function(database_function))), // function implementation
        );

        let sql_results = ctx.sql(&format!("select database()"))?.collect().await?;

        let sql_result_col = array_value_to_string(sql_results[0].column(0), 0).unwrap();
        assert_eq!(sql_result_col, expected_name);

        Ok(())
    }
}

#[tokio::test]
async fn test_func_provider_results() -> Result<()> {
    let p = TestProvider::try_new("test")?;
    p.test("test".to_string());
    Ok(())
}

```

**Comment by wqc200** [comment]
@alamb I want to change the db name, but database() get old db name.

```
use std::sync::Arc;

use arrow::array::ArrayRef;
use arrow::array::StringArray;
use arrow::datatypes::DataType;
use arrow::util::display::array_value_to_string;

use datafusion::error::Result;
use datafusion::execution::context::ExecutionContext;
use datafusion::logical_plan::create_udf;
use datafusion::physical_plan::functions::make_scalar_function;

pub struct TestProvider {
    db_name: String,
    ctx: ExecutionContext,
}

impl TestProvider {
    pub fn try_new(db_name: &str) -> Result<Self> {
        let ctx = ExecutionContext::new();

        Ok(Self {
            db_name: db_name.to_string(),
            ctx,
        })
    }

    pub fn change_db_name(&mut self, db_name: &str) {
        self.db_name = db_name.to_string();
    }

    pub fn register_udf(&mut self) {
        // implementation of `database()` function that returns expected_name
        let captured_name = self.db_name.clone();
        let database_function = move |_args: &[ArrayRef]| {
            let db_name = captured_name.as_str();
            let res = StringArray::from(vec![Some(db_name)]);
            Ok(Arc::new(res) as ArrayRef)
        };

        self.ctx.register_udf(create_udf(
            "database", // fuenction name
            vec![], // input argument types
            Arc::new(DataType::Utf8), // output type
            make_scalar_function(database_function)), // function implementation
        );
    }

    pub async fn get_db_name(&mut self) -> Result<String> {
        let sql_results = self.ctx.sql("select database()")?.collect().await?;

        let sql_result_col = array_value_to_string(sql_results[0].column(0), 0).unwrap();

        Ok(sql_result_col)
    }
}

#[tokio::test]
async fn test_func_provider_results() -> Result<()> {
    let mut p = TestProvider::try_new("test1")?;
    p.register_udf();
    p.change_db_name("test");

    let db_name = p.get_db_name().await?;
    let expected_db_name = "test";

    assert_eq!(db_name, expected_db_name);

    Ok(())
}
```

**Comment by alamb** [comment]
@wqc200  here is a way to use `std::sync::Mutex` to accomplish changing database name:

```
use std::sync::{Arc, Mutex};

use arrow::array::ArrayRef;
use arrow::array::StringArray;
use arrow::datatypes::DataType;
use arrow::util::display::array_value_to_string;

use datafusion::error::Result;
use datafusion::execution::context::ExecutionContext;
use datafusion::logical_plan::create_udf;
use datafusion::physical_plan::functions::make_scalar_function;

pub struct TestProvider {
    // Wrap the db_name in a mutex so it can be changed after registration
    db_name: Arc<Mutex<String>>,
    ctx: ExecutionContext,
}

impl TestProvider {
    pub fn try_new(db_name: &str) -> Result<Self> {
        let ctx = ExecutionContext::new();

        Ok(Self {
            db_name: Arc::new(Mutex::new(db_name.to_string())),
            ctx,
        })
    }

    pub fn change_db_name(&mut self, db_name: &str) {
        *self.db_name.lock().expect("mutex poisoned") = db_name.to_string();
    }

    pub fn register_udf(&mut self) {
        // implementation of `database()` function that returns the
        // current value of `self.db_name`
        let captured_name = self.db_name.clone();
        let database_function = move |_args: &[ArrayRef]| {
            // Lock the mutex, and read current db_name
            let captured_name = captured_name.lock().expect("mutex posioned");
            let db_name = captured_name.as_str();
            let res = StringArray::from(vec![Some(db_name)]);
            Ok(Arc::new(res) as ArrayRef)
        };

        self.ctx.register_udf(create_udf(
            "database", // function name
            vec![], // input argument types
            Arc::new(DataType::Utf8), // output type
            make_scalar_function(database_function)), // function implementation
        );
    }

    pub async fn get_db_name(&mut self) -> Result<String> {
        let sql_results = self.ctx.sql("select database()")?.collect().await?;

        let sql_result_col = array_value_to_string(sql_results[0].column(0), 0).unwrap();

        Ok(sql_result_col)
    }
}

#[tokio::test]
async fn test_func_provider_results() -> Result<()> {
    let mut p = TestProvider::try_new("test1")?;
    p.register_udf();
    p.change_db_name("test");

    let db_name = p.get_db_name().await?;
    let expected_db_name = "test";

    assert_eq!(db_name, expected_db_name);

    Ok(())
}

```

**Pull request closed without merging**